### DISCUSSION TITLE
? Quantifier not matching in 'matchStrings'

### DISCUSSION THREAD

**Discussion by ckstevenson** [renovatebot/renovate#18208] [Request Help] [answered]
Maybe a bug, but often I am the bug. I have set up 3 files, namely `1.yml`, `2.yml`, `3.yml`. They contain:

- `terraform_version: 1.0.5`
- `terraform_version: "1.0.5"`
- `terraform_version: '1.0.5'`
respectively.

When I try
```
      "matchStrings": [
        "terraform_version:\\s*\"?\'?(?<currentValue>.*)\"?\'?\\s*"
      ],
```

only the string with no quotes matches (`terraform_version: 1.0.5`). If I remove the quantifier `?`, the other strings match. From my understanding `\"?\'?` should match 0 or 1 double or single quote, so it should match all cases, but instead it only matches when 0 of the preceding character exists. I verified this should work via regex101, and found examples showing it should work in the docs. Anyone able to tell me what I am missing?

Thanks!

**Comment by RahulGautamSingh** [comment, marked as answer]
Hey @ckstevenson , I am looking into it in the meanwhile it would be helpful if you can create a[ minimal reporoduction](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) so that I dont miss anything while imitating the behaviour.

**Reply by ckstevenson** [reply]
@RahulGautamSingh yep yep, will do that now

**Reply by ckstevenson** [reply]
@RahulGautamSingh https://github.com/ckstevenson/renovate-minimal

**Reply by RahulGautamSingh** [reply]
Github already detects some issue with the string :)

![image](https://user-images.githubusercontent.com/76607510/194708447-a4fe7001-cfae-4143-886c-39da3bce6077.png)



You will need to escape the `'` with 2 back slashes instead of one, like this:
```  "terraform_version:\\s*\"?\\'?(?<currentValue>.*)\"?\\'?\\s*"  ```

After using backslashes I get this result, seems like all the files and their deps are detected 
```log
 INFO: Dependency extraction complete (repository=RahulGautamSingh-testing/renovate-minimal, baseBranch=main)
       "stats": {
         "managers": {
           "github-actions": {"fileCount": 1, "depCount": 2},
           "regex": {"fileCount": 3, "depCount": 3}
         },
         "total": {"fileCount": 4, "depCount": 5}
       }
```

**Reply by ckstevenson** [reply]
Ah I see the same when viewing on GitHub. Seems I need a better linter! I now have another issue when adding the double backslash:

```
           {
             "packageFile": "2.yml",
             "deps": [
               {
                 "depName": "hashicorp/terraform",
                 "currentValue": "1.0.5\"",
                 "datasource": "github-releases",
                 "replaceString": "terraform_version: \"1.0.5\"\n",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "semver",
                 "skipReason": "invalid-value"
               }
             ],
             "matchStrings": [
               "terraform_version:\\s*\"?\\'?(?<currentValue>.*)\"?\\'?\\s*"
             ],
             "depNameTemplate": "hashicorp/terraform",
             "datasourceTemplate": "github-releases"
           },
           {
             "packageFile": "3.yml",
             "deps": [
               {
                 "depName": "hashicorp/terraform",
                 "currentValue": "1.0.5'",
                 "datasource": "github-releases",
                 "replaceString": "terraform_version: '1.0.5'\n",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "semver",
                 "skipReason": "invalid-value"
               }
             ],
             "matchStrings": [
               "terraform_version:\\s*\"?\\'?(?<currentValue>.*)\"?\\'?\\s*"
             ],
             "depNameTemplate": "hashicorp/terraform",
             "datasourceTemplate": "github-releases"
           }
```
any idea why it is now throwing `invalid-value` and matching with `"currentValue": "1.0.5\"",` ?

**Reply by viceice** [reply]
remove the escape for `'`, its not needed

**Reply by viceice** [reply]
`terraform_version:\\s*['\"]?(?<currentValue>.+?)['\"]?\\s*`

use non-greedy current value match

**Reply by ckstevenson** [reply]
> `terraform_version:\\s*['\"]?(?<currentValue>.+?)['\"]?\\s*`
> 
> use non-greedy current value match

Ah I see. I will play with this more later and report back. Thanks!

**Reply by ckstevenson** [reply]
The `?` at the end of the second `['\"]?` causes this regex to only match the first character, i.e. `"currentValue": "1",`. Removing it matched the use cases `terraform_version: "v1.0.5"` and `terraform_version: 'v1.0.5'`, which is fine since I will ensure all values are quoted. I am leaving it as `"terraform_version:\\s*['\"]?(?<currentValue>.+?)['\"]\\s*"`.

Thanks a lot for the help @viceice and @RahulGautamSingh . I truly appreciate it.
